main.cpp

```diff
@@ -288,436 +236,179 @@ struct HeapA
-    DoubleType& multiply(double rhs);
-    DoubleType& divide(double rhs);
-
-    DoubleType& pow(double rhs);
-    DoubleType& pow(const FloatType& rhs);
-    DoubleType& pow(const DoubleType& rhs);
-    DoubleType& pow(const IntType& rhs);
-
-    DoubleType& apply(std::function<DoubleType&(double&)> f);
-    DoubleType& apply(void (*f)(double&));
-
-private:
-    DoubleType& powInternal(double exp);
-
-    double* value = nullptr;
-};
-// ============================================================
-struct IntType
-{
-    explicit IntType(int int_) : value(new int(int_)) { }
-    ~IntType() { delete value; }
-    operator int() const;
-    IntType& operator+=(int rhs);
-    IntType& operator-=(int rhs);
-    IntType& operator*=(int rhs);
-    IntType& operator/=(int rhs);
-
-    IntType& add(int rhs);
-    IntType& subtract(int rhs);
-    IntType& multiply(int rhs);
-    IntType& divide(int rhs);
-
-    IntType& pow(int rhs);
-    IntType& pow(const FloatType& rhs);
-    IntType& pow(const DoubleType& rhs);
-    IntType& pow(const IntType& rhs);
-
-    IntType& apply(std::function<IntType&(int&)> f);
-    IntType& apply(void (*f)(int&));
-
-private:
-    IntType& powInternal(int exp);
-
-    int* value = nullptr;
-};
-// ======================================== Float primitive ==========
-FloatType& FloatType::add(float rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-FloatType& FloatType::subtract(float rhs)
+#include <memory>
+#include <type_traits>
+#include <limits>
+//====================== Primary Template Numeric =====================
+template <typename NumType>
+struct Numeric
 {
-    *value -= rhs;
-    return *this;
-}
+    using Type = NumType;
 
-FloatType& FloatType::multiply(float rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-FloatType& FloatType::divide(float rhs)
-{
-    if (rhs == 0.0f)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-    *value /= rhs;
-    return *this;
-}
-// ======================================== Float operators ===========
-FloatType::operator float() const { return *value; }
-
-FloatType& FloatType::operator+=(float rhs)
-{
-    *value += rhs;
-    return *this;
-}
+    explicit Numeric(Type type_) : value(std::make_unique<Type>(type_)) { }
+    ~Numeric() { value.reset(nullptr); }
+    //========== operators ==========
+    operator Type() const { return *value; }
 
-FloatType& FloatType::operator-=(float rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator*=(float rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-FloatType& FloatType::operator/=(float rhs)
-{
-    if (rhs == 0.0f)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-    *value /= rhs;
-    return *this;
-}
-// ======================================== Float Apply func ==========
-FloatType& FloatType::apply(std::function<FloatType&(float&)> f)
-{
-    if (f)
+    Numeric<Type>& operator+=(Type rhs)
```

Comment:
Inside the templated class, you do not need to write `Numeric<Type>&` as the return type.
`Numeric&` is all you need to write.
- [ ] revise